src/dependency-manager.ts

```diff
@@ -87,4 +104,103 @@ export class DependencyManager {
         //     console.debug('error:' + err)
         // })
     }
+
+    private async downloadDepsRecurisively(rootPath: string, useNpm: boolean): Promise<void> {
+        try {
+            const names = await fs.readdir(rootPath);
+            for (const name of names) {
+                const childPath = path.resolve(rootPath, name);
+                const stat = await fs.stat(childPath);
+                if (stat.isDirectory()) {
+                    if (name === 'node_modules') {
+                        return;
+                    }
+                    const packageFile = path.resolve(childPath, 'package.json');
+                    const lockFile = path.resolve(childPath, 'yarn.lock');
+
+                    if (fs.existsSync(packageFile)) {
+                        if (useNpm) {
+                            this.runNpm(childPath);
+                        } else { // use npm
+                            if (fs.existsSync(lockFile)) {
+                                await this.downloadDeps(childPath);
+                            } else {
+                                this.logger.info('No yarn.lock file existed under: ' + childPath + ', consider enabled npm downloading.');
+                            }
+                        }
+                    } else {
+                        await this.downloadDepsRecurisively(childPath, useNpm);
+                    }
+                }
+            }
+        } catch (e) {
+            this.logger.error(e);
+        }
+    }
+
+    public async downloadFile(url: URL, dest: string): Promise<void> {
+        return new Promise((resolve, reject) => {
+            const lib = url.protocol.startsWith('https') ? https : http;
+            const request = lib.get(url, (response: http.IncomingMessage) => {
+                if (!response.statusCode || response.statusCode < 200 || response.statusCode > 299) {
+                    reject(new Error('Failed to load page, status code: ' + response.statusCode));
+                }
+
+                const extractorStream: internal.Writable = gunzip();
+                const untarStream = tarFs.extract(dest, {
+                    // @ts-ignore
+                    strip: 1,
```

Comment:
Probably want to make sure we don't extract symlinks here, and as always that paths are checked for `../`

Please look at these resources:

- https://snyk.io/blog/severe-security-vulnerability-in-bowers-zip-archive-extraction 
- https://snyk.io/research/zip-slip-vulnerability
- https://hackerone.com/reports/473811

Reply:
I'm guessing tarFs will err out if the stream isn't a valid tar archive.

If not, what happens if http://169.254.169.254/latest/user-data is provided as a package URL?

Reply:
done, and it's can't be an arbitrary file, it will just error out